Placement match providers

KeithRaven opened the pull request:
This is a port of placement match providers from O1 by Carl Woodhouse (see commit a992c1e74edadc11bf3a9e887bfe1a154369d727).

It includes match providers for ContentTypes and ContentParts.

I've updated the readme so show syntax.  Whilst I've shown example of both it may be confusing showing them together as you wouldn't use them together.

It is worth noting that it also includes a correction to a typo in ShapeAlterationBuilder.cs which was required for this.

Thanks k

sebastienros commented:
Great idea. I think I remember that @carlwoodhouse is the one who implemented the extensibility in O1. I will review it soon.

Does it fulfill your needs at least?

sebastienros commented on src/OrchardCore/OrchardCore.DisplayManagement/README.md:
```diff
@@ -44,6 +50,12 @@ Placement information consists of:
 		"display-type": "Detail",
 		"differentiator": "Article-MyTextField",
 
+        "match": {
```
I don't think `match` is necessary. Every filter can be a the root of the node, like `display-type` or `differentiator`.

I would also accept a string, and array of string, of a complex object (for `matchType`). You could even simplify it by not taking `matchType` at all, as i assume we'll always want 'any'. I can't find an example that someone would want to match on having multiple parts on the content item. 

ContentType can also accept multiple values in this case, right?

KeithRaven replied:
I like the idea of loosing the  "match" node in that the matches are doing the same kind of thing (i.e. filtering) as display-type and differentiator as so they would be better together.

That said I personally find it a little confusing having the filters and alterations (place,wrappers, alternates, shape) alongside each other with nothing to distinguish them but I guess it keeps it simpler.

Yes allowing string, array string or object would be a good idea as well.

> I can't find an example that someone would want to match on having multiple parts on the content item.

I haven't got a use case at the minute but it feels like there one out there somewhere :)

sebastienros replied:
Maybe confusing, but only because you are coming from O1, and as long as it's document it should not be an issue.

carlwoodhouse commented:
Really nice to see someone port this to core :) good job, i agree with @sebastienros that i cant really imagine a use case for the matchtype

KeithRaven commented:
Should I take matchType out completely or just from the docs? (I'm guessing completely would be prefered)

sebastienros commented:
Yes please remove it.

KeithRaven commented:
I've made some updates:

- Removed MatchType
- Removed the 'Match' node from placement.json (all unmatched nodes in  json are are deserialized into a filters dictionary in PlacementNode using JsonExtensionData attribute)
- Updated filters to take string or array
- Renamed classes to better fit with exisiting OC naming
- Added path filter provider

I've left the DisplayType check within ShapePlacementStrategy.cs, should this be extracted out as in the O1 version as well or left?

https://github.com/KeithRaven/OrchardCore/blob/9a67d5250b1e1fc04bebc42ae72c32b5309ad1db/src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/ShapePlacementStrategy/ShapePlacementParsingStrategy.cs#L114

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Contents/Placement/ContentPlacementNodeFilterProviders.cs:
```diff
@@ -0,0 +1,99 @@
+                {
+                    var prefix = ct.Substring(0, ct.Length - 1);
+
+                    return (contentItem.ContentType ?? "").StartsWith(prefix)  || (GetStereotype(context) ?? "").StartsWith(prefix);
```
All string comparisons should use `Ordinal` or `OrdinalIgnoreCase`.

sebastienros commented on src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/ShapePlacementStrategy/PathPlacementNodeFilterProvider.cs:
```diff
@@ -0,0 +1,63 @@
+
+            return paths.Any(p => {
+
+                var normalizedPath = p.StartsWith("~/")
```
Can't decipher that, too complex. I'd prefer some ifs in this case.

sebastienros commented:
Some comments then it will be fine. Thanks.

sebastienros commented:
Kebab looks cool, but camel case is the best approach. Please change the other ones then.
